src/validation/validation.cpp

```diff
@@ -2224,9 +2224,7 @@ bool ConnectBlockDependencyOrdering(const CBlock &block,
                 {
                     for (size_t j = 0; j < tx.vin.size(); j++)
                     {
-                        CoinAccessor coin(view, tx.vin[j].prevout);
-                        // isSpend is true for empty coin object (coinEmpty)
-                        if (coin->IsSpent())
```

Comment:
since the coinaccessor lock isn't taken, the new if statement does not guarantee !coin->IsSpent() (some other thread may have affected it).  You'll have to do something like this:

bool abort = false;
do
{
CoinAccessor coin(view, tx.vin[j].prevout);
if (coin->IsSpent) { abort = true; break; }
prevheights[j] = coin->nHeight;
nFees = nFees + coin->out.nValue;
} while(0);

if (abort)
{
if (PV->ChainWorkHasChanged(nStartingChainWork)
...
}

Reply:
im not sure i understand why, we only started to take coinaccessor like this in a recent PR i made which slightly improved performance. compare this PR to the release branch and the logic is the same WRT the have inputs

Reply:
after a short discussion, changes have been made to fix this issue.